src/main/java/org/elasticsearch/action/update/UpdateHelper.java

```diff
@@ -118,6 +118,8 @@ public Result prepare(UpdateRequest request, IndexShard indexShard) {
                     return new Result(update, Operation.NONE, upsertDoc, XContentType.JSON);
                 }
                 indexRequest.source((Map)ctx.get("_source"));
+                indexRequest.routing(request.routing());
+                indexRequest.parent(request.parent());
```

Comment:
I believe the above two lines need to be moved outside of the `if` statement they are in (`if (request.scriptedUpsert() && (request.script() != null)) `), as currently the parent and routing will only be set if it is scripted upset, and not if it is a vanilla one.